### PR TITLE
Move Windows Mixed Reality and OpenVR providers to separate assemblies

### LINKED ISSUE #3074: vNEXT Task: Refactor Controller Mapping Profile to accept individual controller scriptable objects.
# Overview
At the moment each controller is pre-populated in the controller mapping profile. Ideally we want to give developers full control over the devices and controllers they'd like to use. I'd like to propose changing the controller mapping profile from buttons to Scriptable Object slots for maximum flexibility and customization. This will enable us to easily extend and add new controller types in the future. At present it's a bit difficult. (I had plans to do this before beta but was limited by time and higher priority items).

# Breaking Changes
The controller mapping profiles array of mappings will be turned into their own individual scriptable object profiles which will be assigned in the inspector of the controller mapping profile.

# Requirements
* Controller Mapping Profile should have controller object slots to fill instead of buttons.
* Adding new controller implementations should be simple and easy.
* The only UX change should be the buttons be replaced with scriptable object fields.

# Acceptance Criteria
- [x] Users can create and assign individual controller mapping profiles.
- [x] Users can assign actions to specific buttons and inputs using the existing controller popout windows.

### LINKED ISSUE #3357: Create separate WindowsMixedReality provider assembly
Continuation of #3341, moving the WindowsMixedReality provider into the MixedRealityToolkit.Providers folder and creating the appropriate asmdef file.

Approved by shiproom on 2019-01-18, pending completion of #3074.

Blocked by #3074

### LINKED ISSUE #3358: Create separate OpenVR provider assembly
Continuation of #3341, moving the OpenVR provider into the MixedRealityToolkit.Providers folder and creating the appropriate asmdef file.

Approved by shiproom on 2019-01-18, pending completion of #3074.

Blocked by #3074

### PR DESCRIPTION
**Pull request by davidkline-ms**
Overview
---
This change updates how the controller mapping profile and support code works to remove hard coded references to concrete class implementations (thanks Daniel Hofmann for the slick implementation!). The hard coded references was blocking being able to move platform specific implementations out of the core assembly and into redistributable packages.

This formally moves Windows Mixed Reality and OpenVR platform support components into separate assemblies.

As the Unity input provider is not platform specific, it was left in the Core assembly.

Other related changes:
* The Devices folder was renamed to Providers
* ...Core.Devices namespace was changed to ...Core.Providers

NOTE: You MUST delete your Library folder and update any custom profiles upon pulling this change,

Fixes #3074 
Fixes #3357 
Fixes #3358

### CODE REVIEW

**Comment by StephenHodgson** [comment]
> As the Unity input provider is not platform specific, it was left in the Core assembly.

Same could be said about OpenVR...

**Comment by keveleigh** [comment]
>Same could be said about OpenVR...

You need to download the OpenVR package in Unity, while generic Unity input is built in.

**Comment by StephenHodgson** [comment]
That's not true, it's utilizing the generic unity input.

**Comment by keveleigh** [comment]
>That's not true, it's utilizing the generic unity input.

Sure, but you can't run on OpenVR unless you download the package and enable it in settings. You have to take an explicit step to support it.

**Comment by StephenHodgson** [comment]
That doesn't change the fact that the code in this section has no direct dependence on the OpenVR package

**Comment by keveleigh** [comment]
>That doesn't change the fact that the code in this section has no direct dependence on the OpenVR package

Sure, but you won't be getting proper data from the code / axes without the package. This split is based on actual support for specific platforms.

The generic Unity input will function on all. OpenVR / WindowsMR will not function as expected unless on their specific runtime / platform, which need to be explicitly enabled and supported.

**Comment by StephenHodgson** [comment]
> Sure, but you won't be getting proper data from the code / axes without the package. 

That's also not entirely true either. There's other ways to get the input than from the Unity Packages and have them supported from the old input manager.

I suppose it doesn't matter, the community lost their say in what happens in this repository anyway.

File: Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset [hunk 1]
```diff
@@ -4,24 +4,23 @@
...
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
   m_GameObject: {fileID: 0}
   m_Enabled: 1
   m_EditorHideFlags: 0
   m_Script: {fileID: 11500000, guid: 1a1cf5fd47e548e6a23bdd3ddcc00cf6, type: 3}
   m_Name: DefaultMixedRealityControllerMappingProfile
   m_EditorClassIdentifier: 
-  isCustomProfile: 0
+  isCustomProfile: 1
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset]: Good catch!

[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit.SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset]:
```suggestion
  isCustomProfile: 0
```

**Pull request merged**